soccer/planning/InterpolatedPath.hpp

```diff
@@ -0,0 +1,147 @@
+#include <Geometry2d/Point.hpp>
+#include <Geometry2d/Segment.hpp>
+#include <Geometry2d/CompositeShape.hpp>
+#include <Configuration.hpp>
+
+namespace Planning
+{
+	/**
+	 * @brief Represents a motion path as a series of {pos, vel} pairs.
+	 *
+	 * @details The path represents a function of position given time that the robot should follow.  A
+	 * line-segment-based path comes from the planner, then we use cubic bezier curves to interpolate
+	 * and smooth it out.  This is done via the evaulate() method.
+	 */
+	class InterpolatedPath: public Path
+	{
+		public:
+
+			/** default path is empty */
+			InterpolatedPath() {}
+
+			/** constructor with a single point */
+			InterpolatedPath(const Geometry2d::Point& p0);
+
+			/** constructor from two points */
+			InterpolatedPath(const Geometry2d::Point& p0, const Geometry2d::Point& p1);
+
+			bool empty() const
+			{
+				return points.empty();
+			}
+
+			void clear()
+			{
+				points.clear();
+			}
+
+			/**
+			 * Calulates the length of the path
+			 *
+			 * \param[in] 	start Index of point in path to use at start point.
+			 * \returns 	the length of the path starting at point (start).
+			 */
+			float length(unsigned int start = 0) const;
+
+			/**
+			 * Calulates the length of the path
+			 *
+			 * \param[in] 	start Index of point in path to use at start point.
+			 * \param[in] 	end Index of point in path to use at end point.
+			 * \returns 	the length of the path starting at point (start) and ending at point [end].
+			 */
+			float length(unsigned int start, unsigned int end) const;
+
+			/** Returns the length of the path from the closet point found to @a pt */
+			float length(const Geometry2d::Point &pt) const;
+
+			/** Returns number of waypoints */
+			size_t size() const { return points.size(); }
+
+			/** returns true if the path has non-zero size */
+			bool valid() const { return !points.empty(); }
+
+			// Returns the index of the point in this path nearest to pt.
+			int nearestIndex(const Geometry2d::Point &pt) const;
+
+			/** returns the nearest segement of @a pt to the path */
+			Geometry2d::Segment nearestSegment(const Geometry2d::Point &pt) const;
+
+			// Returns the shortest distance from this path to the given point
+			float distanceTo(const Geometry2d::Point &pt) const;
+
+			// Returns the start of the path
+			boost::optional<Geometry2d::Point> start() const;
+
+			// Returns a new path starting from a given point
+			void startFrom(const Geometry2d::Point& pt, Planning::InterpolatedPath& result) const;
+
+			//Returns the destination of this path (the last point in the points array)
+			virtual boost::optional<Geometry2d::Point> destination() const;
+
+			// Returns true if the path never touches an obstacle or additionally, when exitObstacles is true, if the path
+			// starts out in an obstacle but leaves and never re-enters any obstacle.
+			virtual bool hit(const Geometry2d::CompositeShape &shape, unsigned int start = 0) const;
```

Comment:
Yeah. That's not right.... looks like it was written in 2009 lol. I'll fix
it. Also TODO:  We should also change collision detection to not use
exceptions at some point.
On Feb 3, 2015 2:44 PM, "Justin Buchanan" <notifications@github.com> wrote:

> In soccer/planning/InterpolatedPath.hpp
> <https://github.com/RoboJackets/robocup-software/pull/194#discussion_r24032932>
> :
>
> > +
> > +			// Returns the shortest distance from this path to the given point
> > +			float distanceTo(const Geometry2d::Point &pt) const;
> > +
> > +			// Returns the start of the path
> > +			boost::optional<Geometry2d::Point> start() const;
> > +
> > +			// Returns a new path starting from a given point
> > +			void startFrom(const Geometry2d::Point& pt, Planning::InterpolatedPath& result) const;
> > +
> > +			//Returns the destination of this path (the last point in the points array)
> > +			virtual boost::optional<Geometry2d::Point> destination() const;
> > +
> > +			// Returns true if the path never touches an obstacle or additionally, when exitObstacles is true, if the path
> > +			// starts out in an obstacle but leaves and never re-enters any obstacle.
> > +			virtual bool hit(const Geometry2d::CompositeShape &shape, unsigned int start = 0) const;
>
> The naming seems a little off here... shouldn't it return true if it finds
> a collision and false if it's collision-free?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/RoboJackets/robocup-software/pull/194/files#r24032932>
> .
>